File: server/src/services/cognee-cloud/init-datasets.ts
/**
 * Initialize all datasets in Cognee Cloud.
 *
 * Creates all 39 datasets (6 shared + 33 tenant) in Cognee Cloud Premium.
 * Run this script once to set up the knowledge graph infrastructure.
 *
 * Usage:
 *   npx tsx src/services/cognee-cloud/init-datasets.ts
 */

import { getCogneeCloudClient } from './client.js';
import { getSharedDatasets, getTenantDatasets, type DatasetDefinition } from './datasets.js';

async function createDataset(datasetDef: DatasetDefinition, retries = 3): Promise<boolean> {
  const client = getCogneeCloudClient();

  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      console.log(`Creating dataset: ${datasetDef.name} (attempt ${attempt}/${retries})`);
      console.log(`  Description: ${datasetDef.description}`);
      console.log(`  Category: ${datasetDef.category}`);
      console.log(`  Public: ${datasetDef.isPublic}`);

      // Add a placeholder document to create the dataset
      const result = await client.addData(
        `# ${datasetDef.name}\n\n${datasetDef.description}\n\nThis is a placeholder document to initialize the dataset.`,
        datasetDef.name,
      );

      console.log(`  ✅ Created: ${result.dataset_id}`);
      console.log(`  Status: ${result.status}`);
      console.log();

      return true;
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);

      if (attempt < retries && errorMsg.includes('Could not refresh instance')) {
        console.log(`  ⚠️  Retry ${attempt}/${retries}: ${errorMsg}`);
        // Wait 2 seconds before retry
        await new Promise((resolve) => setTimeout(resolve, 2000));
        continue;
      }

      console.log(`  ❌ Error: ${errorMsg}`);
      console.log();
      return false;
    }
  }

  return false;
}

async function createAllDatasets(): Promise<number> {
  console.log('='.repeat(80));
  console.log('Cognee Cloud Dataset Initialization');
  console.log('='.repeat(80));
  console.log();

  const sharedDatasets = getSharedDatasets();
  const tenantDatasets = getTenantDatasets();

  console.log(`📊 Total datasets to create: ${sharedDatasets.length + tenantDatasets.length}`);
  console.log(`   - Shared (public): ${sharedDatasets.length}`);
  console.log(`   - Per-tenant: ${tenantDatasets.length}`);
  console.log();

  // Create shared datasets
  console.log('='.repeat(80));
  console.log('SHARED DATASETS (Public Knowledge)');
  console.log('='.repeat(80));
  console.log();

  let sharedSuccess = 0;
  for (const dataset of sharedDatasets) {
    if (await createDataset(dataset)) {
      sharedSuccess++;
    }
    // Small delay between datasets to avoid rate limiting
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }

  console.log(`✅ Shared datasets created: ${sharedSuccess}/${sharedDatasets.length}`);
  console.log();

  // Create tenant datasets
  console.log('='.repeat(80));
  console.log('TENANT DATASETS (Per-User Financial Data)');
  console.log('='.repeat(80));
  console.log();

  let tenantSuccess = 0;
  for (const dataset of tenantDatasets) {
    if (await createDataset(dataset)) {
      tenantSuccess++;
    }
    // Small delay between datasets to avoid rate limiting
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }

  console.log(`✅ Tenant datasets created: ${tenantSuccess}/${tenantDatasets.length}`);
  console.log();

  // Summary
  console.log('='.repeat(80));
  console.log('SUMMARY');
  console.log('='.repeat(80));
  console.log(
    `Total datasets created: ${sharedSuccess + tenantSuccess}/${sharedDatasets.length + tenantDatasets.length}`,
  );
  console.log();

  if (sharedSuccess + tenantSuccess === sharedDatasets.length + tenantDatasets.length) {
    console.log('🎉 All datasets created successfully!');
    return 0;
  } else {
    console.log('⚠️  Some datasets failed to create. Check errors above.');
    return 1;
  }
}

async function testConnection(): Promise<boolean> {
  console.log('Testing connection to Cognee Cloud...');
  console.log();

  try {
    const client = getCogneeCloudClient();

    // Try a simple add operation
    const result = await client.addData('Test connection to Cognee Cloud', 'test_connection');

    console.log('✅ Connection successful!');
    console.log(`   Dataset ID: ${result.dataset_id}`);
    console.log(`   Status: ${result.status}`);
    console.log();

    return true;
  } catch (error) {
    console.log(`❌ Connection failed: ${error instanceof Error ? error.message : String(error)}`);
    console.log();
    return false;
  }
}

async function main(): Promise<number> {
  // Test connection first
  if (!(await testConnection())) {
    console.log('❌ Cannot connect to Cognee Cloud. Check your API key.');
    return 1;
  }

  // Create all datasets
  return await createAllDatasets();
}

// Run if executed directly
main()
  .then((exitCode) => {
    process.exit(exitCode);
  })
  .catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });

export { main, testConnection, createAllDatasets };

File: server/src/services/cognee-cloud/init_datasets.py
#!/usr/bin/env python3
"""
Initialize all datasets in Cognee Cloud.

Creates all 39 datasets (6 shared + 33 tenant) in Cognee Cloud Premium.
Run this script once to set up the knowledge graph infrastructure.

Usage:
    python -m server.src.services.cognee-cloud.init_datasets
"""

import asyncio
import sys
from typing import List
from client import get_cognee_cloud_client
from datasets import get_shared_datasets, get_tenant_datasets, DatasetDefinition


async def create_dataset(client, dataset: DatasetDefinition) -> bool:
    """
    Create a single dataset in Cognee Cloud.
    
    Args:
        client: CogneeCloudClient instance
        dataset: DatasetDefinition to create
        
    Returns:
        True if successful, False otherwise
    """
    try:
        print(f"Creating dataset: {dataset.name}")
        print(f"  Description: {dataset.description}")
        print(f"  Category: {dataset.category}")
        print(f"  Public: {dataset.is_public}")
        
        # Add a placeholder document to create the dataset
        # Cognee Cloud creates datasets on first add() call
        result = await client.add_data(
            data=f"# {dataset.name}\n\n{dataset.description}\n\nThis is a placeholder document to initialize the dataset.",
            dataset_name=dataset.name
        )
        
        print(f"  ✅ Created: {result.dataset_id}")
        print(f"  Status: {result.status}")
        print()
        
        return True
        
    except Exception as e:
        print(f"  ❌ Error: {str(e)}")
        print()
        return False


async def create_all_datasets():
    """Create all datasets in Cognee Cloud."""
    print("=" * 80)
    print("Cognee Cloud Dataset Initialization")
    print("=" * 80)
    print()
    
    client = get_cognee_cloud_client()
    
    # Get all datasets
    shared_datasets = get_shared_datasets()
    tenant_datasets = get_tenant_datasets()
    
    print(f"📊 Total datasets to create: {len(shared_datasets) + len(tenant_datasets)}")
    print(f"   - Shared (public): {len(shared_datasets)}")
    print(f"   - Per-tenant: {len(tenant_datasets)}")
    print()
    
    # Create shared datasets
    print("=" * 80)
    print("SHARED DATASETS (Public Knowledge)")
    print("=" * 80)
    print()
    
    shared_success = 0
    for dataset in shared_datasets:
        if await create_dataset(client, dataset):
            shared_success += 1
    
    print(f"✅ Shared datasets created: {shared_success}/{len(shared_datasets)}")
    print()
    
    # Create tenant datasets (without tenant prefix for now)
    print("=" * 80)
    print("TENANT DATASETS (Per-User Financial Data)")
    print("=" * 80)
    print()
    
    tenant_success = 0
    for dataset in tenant_datasets:
        if await create_dataset(client, dataset):
            tenant_success += 1
    
    print(f"✅ Tenant datasets created: {tenant_success}/{len(tenant_datasets)}")
    print()
    
    # Summary
    print("=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print(f"Total datasets created: {shared_success + tenant_success}/{len(shared_datasets) + len(tenant_datasets)}")
    print()
    
    if shared_success + tenant_success == len(shared_datasets) + len(tenant_datasets):
        print("🎉 All datasets created successfully!")
        return 0
    else:
        print("⚠️  Some datasets failed to create. Check errors above.")
        return 1


async def test_connection():
    """Test connection to Cognee Cloud."""
    print("Testing connection to Cognee Cloud...")
    print()
    
    try:
        client = get_cognee_cloud_client()
        
        # Try a simple add operation
        result = await client.add_data(
            data="Test connection to Cognee Cloud",
            dataset_name="test_connection"
        )
        
        print(f"✅ Connection successful!")
        print(f"   Dataset ID: {result.dataset_id}")
        print(f"   Status: {result.status}")
        print()
        
        return True
        
    except Exception as e:
        print(f"❌ Connection failed: {str(e)}")
        print()
        return False


async def main():
    """Main entry point."""
    # Test connection first
    if not await test_connection():
        print("❌ Cannot connect to Cognee Cloud. Check your API key.")
        return 1
    
    # Create all datasets
    return await create_all_datasets()


if __name__ == "__main__":
    exit_code = asyncio.run(main())
    sys.exit(exit_code)

